src/trap_routines.rs

```diff
@@ -0,0 +1,272 @@
+use crate::{
+    error::VMError,
+    hardware::{Memory, Register, Registers},
+    utils::{getchar, stdout_flush, stdout_write, update_flags},
+};
+use std::{
+    io::{Read, Write},
+    num::TryFromIntError,
+};
+
+const NULL: u16 = 0x0000;
+
+pub enum TrapCode {
+    GetC,
+    Out,
+    Puts,
+    In,
+    PutsP,
+    Halt,
+}
+
+impl TryFrom<u16> for TrapCode {
+    type Error = VMError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0x20 => Ok(TrapCode::GetC),
+            0x21 => Ok(TrapCode::Out),
+            0x22 => Ok(TrapCode::Puts),
+            0x23 => Ok(TrapCode::In),
+            0x24 => Ok(TrapCode::PutsP),
+            0x25 => Ok(TrapCode::Halt),
+            _ => Err(VMError::Conversion(String::from(
+                "Invalid u16 for TrapCode conversion",
+            ))),
+        }
+    }
+}
+
+/// Read one character from the stdin.
+pub fn get_c(regs: &mut Registers, reader: &mut impl Read) -> Result<(), VMError> {
+    let buffer = getchar(reader)?;
+    let char: u16 = buffer[0].into();
+    regs[Register::R0] = char;
+    update_flags(Register::R0, regs);
+    Ok(())
+}
+
+/// Output a single character.
```

Comment:
To keep consistency with [the previous function](https://github.com/DiegoCivi/virtual-machine-LC3/blob/ca2a80a88691dbe2e83a94dd72b3c8d654abe74e/src/trap_routines.rs#L40):

```suggestion
/// Writes a single character into stdout.
```